src/Form/AppSettingsForm.php

```diff
@@ -69,90 +92,70 @@ public function buildForm(array $form, FormStateInterface $form_state) {
       '#default_value' => $generalConfig->get('display_as_select'),
     ];
 
-    $form['api_product']['associate_apps'] = [
-      '#type' => 'checkbox',
-      '#title' => $this->t('Associate apps with API Products'),
-      '#default_value' => $generalConfig->get('associate_apps'),
-    ];
-
     $form['api_product']['user_select'] = [
       '#type' => 'checkbox',
       '#title' => $this->t('Let user select the product(s)'),
       '#default_value' => $generalConfig->get('user_select'),
-      '#states' => [
-        'visible' => [
-          ':input[name="associate_apps"]' => ['checked' => TRUE],
+      '#ajax' => [
+        'callback' => '::apiProductListCallback',
+        'wrapper' => 'default-api-product-multiple',
+        'progress' => [
+          'type' => 'throbber',
+          'message' => '',
         ],
       ],
     ];
 
-    /** @var string[] $default_products */
-    $default_products = $generalConfig->get('default_products') ?: [];
-    $product_list = [];
-    /** @var \Drupal\apigee_edge\Entity\ApiProduct[] $products */
-    try {
-      $products = ApiProduct::loadMultiple();
-      foreach ($products as $product) {
-        $product_list[$product->id()] = $product->getDisplayName();
-      }
-    }
-    catch (EntityStorageException $e) {
-      // Apigee Edge credentials are missing/incorrect or something else went
-      // wrong. Do not redirect the user to the error page.
-      $product_list = [];
-      $this->messenger()->addError($this->t('Unable to retrieve API product list from Apigee Edge. Please ensure that <a href=":link">Apigee Edge connection settings</a> are correct.'), [
-        ':link' => Url::fromRoute('apigee_edge.settings')->toString(),
-      ]);
-    }
+    // It's necessary to add a wrapper, because if the ID is added to the
+    // checkboxes form element, then that will not be properly rendered
+    // (the label is duplicated).
+    $form['api_product']['default_api_product_multiple_container'] = [
+      '#type' => 'container',
+      '#id' => 'default-api-product-multiple',
+    ];
 
-    $form['api_product']['default_api_product_multiple'] = [
+    $form['api_product']['default_api_product_multiple_container']['default_api_product_multiple'] = [
       '#type' => 'checkboxes',
```

Comment:
You could have tried to use multiple checkbox form element if "checkboxes" that problematic :)

Reply:
But I should create custom validation for the multiple checkbox form element approach (must choose at least one product if user_select is unchecked).